lib/recursive-methods.rb

```diff
@@ -1,49 +1,134 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    raise ArgumentError if n < 0
+    if n == 1 || n == 0 
+        return 1
+    else 
+        return n * factorial(n-1)
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2)
+# Space complexity: O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    if s.length == 0
+        return ""
+    else   
+        return s[-1] + reverse(s[0...-1]) 
+    end 
 end
 
-# Time complexity: ?
-# Space complexity: ?
+def reverse_helper(s, first = 0, last = s.length - 1)
+    if first < last 
+        temp = s[first]
+        s[first] = s[last]
+        s[last] = temp
+        return reverse_helper(s, first + 1, last - 1)
+    end 
+    return s
+end 
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+    return reverse_helper(s, 0, s.length - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(n)
+def bunny(bunnies, ears = 0)
+    if bunnies == (ears/2)   
+        return ears 
+    else 
+        return bunny(bunnies, ears + 1)
+    end 
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def nested(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(n)
+def nested(string, first = 0, last = (string.length - 1))
+    if string == "" || first > last #if first index is greater than the last, all () were checked, that is the base case.  
+        return true 
+    elsif (string.length % 2 != 0) 
+        return false 
+    elsif string[first] != "(" || string[last] != ")"
+        return false 
+    else 
+        return nested(string, first + 1, last - 1)
+    end 
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(n)
+def search(array, value, index = 0) 
+    return false if array == []
+    if value == array[index]
+        return true 
+    elsif index == array.length - 1
+        return false
+    else 
+        return search(array, value, index + 1)
+    end 
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def is_palindrome(s)
```

Comment:
The space complexity here is O(n^2) because you're creating a new array with each recursive call.